_transcripts/008-oli-miri.md

```diff
@@ -0,0 +1,417 @@
+actually do some kind of undefined behavior that it cannot detect. Yet.
+
+__Ben__: I want to kind of back up here, too, because it's kind of blowing my
+mind, in some cases. Let's just talk about undefined behavior as a general
+concept, like in C. So in C, part of the thing with undefined behavior is, hey,
+here's a thing that we as a compiler as a language, you can't do legally. It is
+illegal to do these things, but we can't ever tell if you do. At least not
+tractably, or all the time. And so in C, there are things that you just
+shouldn't do, because if you do, you are writing undefined C which is not like
+compliant C, which happens all the time. It's just unavoidable. It's very easy
+to do. In Rust, it should be harder to do. But if you have an `unsafe` block,
+you can still do undefined things. But I think you were telling me that your
+goal is to eventually make it so that Miri can detect 100% of undefined behavior
+in Rust. So what does that even mean? Is it even undefined anymore? If we could
+detect it 100% of the time?
+
+__Oli__: So, we need to remember that Miri runs your program like it were a
+virtual machine, so it will only detect if this specific run had undefined
+behavior. It is not a general, your program is free of undefined behavior, if
+Miri passes. It is, in this run Miri did not find any undefined behavior, and
+except for the two or three edge cases that we know of, every kind of undefined
+behavior would have been detected.
+
+__Ben__: And those edge cases you're saying you have— you're working on making
+those go away?
+
+__Oli__: Right. But still, only for the specific input you gave to this specific
+program run, you would have had detected undefined behavior. If you change the
+input, you would have to re-run again to check if that input would cause
+undefined behavior.
+
+__Ben__: Interesting. So it's more like the sanitizers, kind of, that LLVM
+provides, but tailored for Rust specifically. And I guess hopefully more
+complete. You wouldn't need, like, both an address sanitizer and a thread
+sanitizer, because that should just cover all undefined behavior for any given
+run. Is it much slower, this virtual machine, then? You're laughing. That's a
+good sign, right?
+
+__Oli__: No, it is ridiculously slower. It is several orders of magnitude
+slower.
+
+__Ben__: Okay.
+
+__Oli__: The idea is not to run, like, very large programs in it, but like, to
+run your, for example, your entire test suite in it or something. But only the
+unit tests, like only small, self-contained things. Mostly around your unsafe
+code. You will try— at least cover the unsafe code because the safe code should
+be fine. So you should be running your your tests inside Miri. Especially if the
+source code that is employed has unsafe code.
+
+__Ben__: And how you would do that is, you would use this rustup component which
+you mentioned before. How would you (_unintelligible— 17:20_) "rustup install
```

Comment:
I think he's trying to say "what's the install ...", but then interrupts himself with what follows. Not sure that's useful to include.